giraffe/README.md

```diff
@@ -342,14 +342,24 @@ When using the comma separated values (CSV) from the Flux query as the `fluxResp
 
 ### Data properties
 
-- **table**: _Object. Optional. When missing, `config` will look for **fluxResponse** to create the **table**. When both are missing, `config` will create an empty **table**._ A data object produced by [Giraffe's utility functions](#utility-functions). Houses the data and getters and setters for that data, for the `<Plot>`.
+There are three ways to pass in data to Giraffe depending on the layer (graph) type:
+
+Data Property Name  | SimpleTable            | TableGraph | All Other Layers
+--------------------|------------------------|------------|------------------
+fluxResponse        |  optional              |  required  | optional
+fromFluxResult      |  optional (preferred)  |  not used  | not used
+table               |  not used              |  not used  | optional (preferred)
+
+- **fluxResponse**: _String. Required for TableGraph layer. Optional for all other layers._ [The unparsed string of comma separated values returned from a Flux query representing the data to be visualized](https://github.com/influxdata/flux/blob/master/docs/SPEC.md#csv). For TableGraph, it is required, as it is the only property that represents the data. For SimpleTable, this property is ignored when **fromFluxResult** is present, as **fromFluxResult** takes precendence, For all other layer types, this property is ignored when **table** is present, as **table** takes precendence.
+
+- **fromFluxResult**: _Object. Optional and used only by the SimpleTable layer type._ An object containing the properties of the result of parsing **fluxResponse**. Using this property prevents **SimpleTable** from unnecessarily re-parsing the **fluxResponse**. When both are missing, `config` will create an empty data set without any parsing. Ignored by all other layer types, see below.
+
+- **table**: _Object. Optional._ The data structure for a layer obtained from parsing a **fluxResponse**. **Not used by the SimpleTable and TableGraph** layer types (SimpleTable and TableGraph, see above). When present, overrides **fluxResponse**. When missing, `config` will look for **fluxResponse** and parse it to create the **table**. When both are missing, `config` will create an empty **table** without any parsing. This property is a data object produced by [Giraffe's utility functions](#utility-functions). Houses the data and getters and setters for that data, for the `<Plot>`.
```

Comment:
great job doing the work to understand this enough to summarize it here. The different types of flux data aren't at all obvious when looking at this from a high level.

After reading the summary in this form, the obvious question is: all this complexity in different types and optional responses and different preferences depending on which response is being used necessary, or is it just incidental? Can we eventually move to a state where we have a single response type and consistent interface?

Reply:
Incidental.

Giraffe was originally part of the UI application. As we have slowly moved the visualization responsibilities from UI to Giraffe, the interface between them has evolved. On top of that, new visualization types were added during this transition, complicating the interface further.

The different ways of passing data to Giraffe should be viewed as flexibility rather than inconsistency. We were supporting open source users of Giraffe who write their consuming UI application differently than our own UI, and of course our own UI. The different interfaces exist to keep backwards compatibility.